README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/saved-request-detail.svg)](https://www.npmjs.com/package/@advanced-rest-client/saved-request-detail)

[![Build Status](https://travis-ci.org/advanced-rest-client/saved-request-detail.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/saved-request-detail)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/advanced-rest-client/saved-request-detail)


# &lt;saved-request-detail&gt;

Details applet for saved request object.

## Example:

```html
<saved-request-detail
  request='{"method":"GET","name":"Demo request","legacyProject":"some-id","url":"https://api.domain.com/endpoint","created":1532573782247}'
></saved-request-detail>
```

## Usage

### Installation
```
npm install --save @advanced-rest-client/saved-request-detail
```

### In a LitElement

```js
import { LitElement, html } from 'lit-element';
import './node_modules/@advanced-rest-client/saved-request-detail/saved-request-detail.js';

class SampleElement extends LitElement {
  render() {
    return html`
    <saved-request-detail
      .request="${this.request}"
      @edit-request="${this._onEdit}"
      @delete-request="${this._onDelete}"
      @navigate="${this._onNavigate}"></saved-request-detail>
    `;
  }

  _onEdit() {
    ...
  }

  _onDelete() {
    ...
  }

  _onNavigate() {
    ...
  }
}
customElements.define('sample-element', SampleElement);
```

## Development

```sh
git clone https://github.com/advanced-rest-client/saved-request-detail
cd saved-request-detail
npm install
```

### Running the tests

```sh
npm test
```

## API components

This components is a part of [API components ecosystem](https://elements.advancedrestclient.com/)


CHANGELOG.md
<a name="2.0.0"></a>
# [2.0.0](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...2.0.0) (2019-01-04)


### Docs

* Updating documentation ([cd70fe5e5589438364d203076c8d2f349c910a2b](https://github.com/advanced-rest-client/saved-request-detail/commit/cd70fe5e5589438364d203076c8d2f349c910a2b))

### Fix

* Fixed misspels ([eef3ca8c568c8e5b5b7a1388cfad86d32bd7ffd6](https://github.com/advanced-rest-client/saved-request-detail/commit/eef3ca8c568c8e5b5b7a1388cfad86d32bd7ffd6))

### New

* Adding `isSaved` computed property ([aa9634af9d8e01037d0f71bfd26fef8f75e57f21](https://github.com/advanced-rest-client/saved-request-detail/commit/aa9634af9d8e01037d0f71bfd26fef8f75e57f21))

### Update

* Adding updated time label ([401f9e6502a6768951d8956b0906e38ca645e3ad](https://github.com/advanced-rest-client/saved-request-detail/commit/401f9e6502a6768951d8956b0906e38ca645e3ad))
* Redesign of project link action ([a0516e67c23830a48a0ae63c78751adf6fdc6329](https://github.com/advanced-rest-client/saved-request-detail/commit/a0516e67c23830a48a0ae63c78751adf6fdc6329))
* Updating component after model change ([d94d6211af86622be4e1c88270c9f3c395a84153](https://github.com/advanced-rest-client/saved-request-detail/commit/d94d6211af86622be4e1c88270c9f3c395a84153))
* Updating styling API ([37894552e75f0bf7fafa001fa2f3fc8ec8f43eb0](https://github.com/advanced-rest-client/saved-request-detail/commit/37894552e75f0bf7fafa001fa2f3fc8ec8f43eb0))
* Updating tests ([02684e6942bf8d79ba1fd368e5725ad6f2e1c26a](https://github.com/advanced-rest-client/saved-request-detail/commit/02684e6942bf8d79ba1fd368e5725ad6f2e1c26a))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...0.1.3) (2017-10-03)




<a name="0.1.2"></a>
## [0.1.2](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.1...0.1.2) (2017-10-03)


### New

* Added markdown support for description ([7c658184f68ad818bf69e2ba5b0021baa01a6103](https://github.com/advanced-rest-client/saved-request-detail/commit/7c658184f68ad818bf69e2ba5b0021baa01a6103))



<a name="0.1.1"></a>
## 0.1.1 (2017-09-22)


### Update

* Added sauce configuration for tests ([45eadb3daab1f07096c6ffc338cb30e21edfd9dd](https://github.com/advanced-rest-client/saved-request-detail/commit/45eadb3daab1f07096c6ffc338cb30e21edfd9dd))



# [3.0.0](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...3.0.0) (2019-09-06)



## [3.0.1](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...3.0.1) (2019-09-06)



## [3.0.2](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...3.0.2) (2019-09-30)



## [3.0.3](https://github.com/advanced-rest-client/saved-request-detail/compare/0.1.2...3.0.3) (2019-10-08)





test/saved-request-detail.test.js
import { fixture, assert, html } from '@open-wc/testing';
import * as sinon from 'sinon/pkg/sinon-esm.js';
import { DataGenerator } from '@advanced-rest-client/arc-data-generator/arc-data-generator.js';
// import * as MockInteractions from '@polymer/iron-test-helpers/mock-interactions.js';
import '../saved-request-detail.js';

describe('<saved-request-detail>', function() {
  async function projectFixture(request, ids) {
    if (ids) {
      request.projects = ids;
    }
    const element = await fixture(html`<saved-request-detail></saved-request-detail>`);
    element._request = request;
    await element._requestChanged(request);
    return element;
  }

  async function basicFixture() {
    return await fixture(html`<saved-request-detail></saved-request-detail>`);
  }

  let project;
  before(async () => {
    const data = await DataGenerator.insertProjectsData({
      projectsSize: 1
    });
    project = data[0];
  });

  after(async () => {
    await DataGenerator.destroySavedRequestData();
  });

  describe('basics', () => {
    it('reads project data', async () => {
      const request = DataGenerator.generateSavedItem({});
      const element = await projectFixture(request, [project._id]);
      assert.typeOf(element._projects, 'array');
      const p = element._projects[0];
      assert.typeOf(p.name, 'string', 'Project has name');
      assert.isFalse(p.missing, 'Project is not missing');
    });

    it('reads legacy project data', async () => {
      const request = DataGenerator.generateSavedItem({});
      request.legacyProject = project._id;
      const element = await projectFixture(request);
      assert.typeOf(element._projects, 'array');
    });

    it('Keeps non exiting project on the list', async () => {
      const request = DataGenerator.generateSavedItem({});
      const element = await projectFixture(request, [project._id, 'non-exiting']);
      assert.typeOf(element._projects, 'array', 'Projects is computed');
      assert.lengthOf(element._projects, 2, 'Has 2 projects on the list');
      const p = element._projects[1];
      assert.equal(p.name, 'non-exiting', 'Non exiting project has name');
      assert.isTrue(p.missing, 'Non exiting project is missing');
    });

    it('Fires navigate event for project', async () => {
      const request = DataGenerator.generateSavedItem({});
      const element = await projectFixture(request, [project._id]);
      const node = element.shadowRoot.querySelector('[data-action="open-project"]');
      const spy = sinon.spy();
      element.addEventListener('navigate', spy);
      assert.ok(node, 'Has project open button');
      node.click();
      assert.isTrue(spy.called);
    });

    it('Renders delete button when saved', async () => {
      const request = DataGenerator.generateSavedItem({});
      request._rev = 'test';
      const element = await projectFixture(request, [project._id]);
      const node = element.shadowRoot.querySelector('[data-action="delete-request"]');
      assert.ok(node);
    });

    it('Fires delete-request event', async () => {
      const request = DataGenerator.generateSavedItem({});
      request._rev = 'test';
      const element = await projectFixture(request, [project._id]);
      let eventData;
      element.addEventListener('delete-request', function(e) {
        eventData = e.detail;
        assert.isFalse(e.bubbles);
      });
      const node = element.shadowRoot.querySelector('[data-action="delete-request"]');
      node.click();
      assert.typeOf(eventData, 'object');
      assert.typeOf(eventData.item, 'object');
    });

    it('Fires edit-request event', async () => {
      const request = DataGenerator.generateSavedItem({});
      request._rev = 'test';
      const element = await projectFixture(request, [project._id]);
      let eventData;
      element.addEventListener('edit-request', function(e) {
        eventData = e.detail;
        assert.isFalse(e.bubbles);
      });
      const node = element.shadowRoot.querySelector('[data-action="edit-request"]');
      node.click();
      assert.typeOf(eventData, 'object');
      assert.typeOf(eventData.item, 'object');
    });
  });

  describe('_processProjectsResponse()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Returns undefined when no argument', () => {
      const result = element._processProjectsResponse();
      assert.isUndefined(result);
    });

    it('Returns undefined when empty array', () => {
      const result = element._processProjectsResponse([]);
      assert.isUndefined(result);
    });

    it('Returns array of items', () => {
      const result = element._processProjectsResponse([{ _id: 'test' }]);
      assert.typeOf(result, 'array');
      assert.lengthOf(result, 1);
    });

    it('Adds "missing" property with false to an existing item', () => {
      const result = element._processProjectsResponse([{}]);
      assert.isFalse(result[0].missing);
    });

    it('Adds "missing" property with true to empty item', () => {
      const result = element._processProjectsResponse([undefined], ['test']);
      assert.isTrue(result[0].missing);
    });

    it('Adds "name" property to empty item', () => {
      const result = element._processProjectsResponse([undefined], ['test']);
      assert.equal(result[0].name, 'test');
    });
  });

  describe('_readProjects()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Sets `_projects` to `undefined` when no argument', async () => {
      element._projects = [];
      await element._readProjects();
      assert.isUndefined(element._projects);
    });

    it('Sets _projects property', async () => {
      await element._readProjects([project._id])
      assert.typeOf(element._projects, 'array');
      assert.lengthOf(element._projects, 1);
    });
  });

  describe('_requestChanged()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Sets `_projects` to `undefined` when no argument', async () => {
      element._projects = [];
      await element._requestChanged();
      assert.isUndefined(element._projects);
    });

    it('reads project data', async () => {
      await element._requestChanged({
        projects: [project._id]
      })
      assert.typeOf(element._projects, 'array');
      assert.lengthOf(element._projects, 1);
    });
  });
});


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('deepmerge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'saved-request-detail';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    'SL_Safari',
    // 'SL_EDGE'
  ];
  if (process.env.TRAVIS) {
    const buildLabel = 'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


saved-request-detail.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   saved-request-detail.js
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.
// tslint:disable:no-any describes the API as best we are able today

import {LitElement, html, css} from 'lit-element';

import {ArcResizableMixin} from '@advanced-rest-client/arc-resizable-mixin/arc-resizable-mixin.js';

declare namespace UiElements {

  /**
   * Details applet for saved request object.
   *
   * The applet doesn't support data edit. Element / app hosting this element
   * must handle events sent by this element and support edit action.
   *
   * ### Example
   *
   * ```html
   * <saved-request-detail request="{...}"></saved-request-detail>
   * ```
   *
   * ### Styling
   *
   * `<saved-request-detail>` provides the following custom properties and
   * mixins for styling:
   *
   * Custom property | Description | Default
   * ----------------|-------------|----------
   * `--saved-request-detail-color` | Color of the element | `inherit`
   * `--saved-request-detail-background-color` | Background color of the element | `inherit`
   * `--saved-request-detail-padding` | Padding of the element. | ``
   * `--saved-request-detail-description-max-width` | Max width of the description element | `700px`
   * `--saved-request-detail-description-color` | Color of the request description | `rgba(0, 0, 0, 0.64)`
   * `--saved-request-detail-action-icon-color` | Color of the icon in the action button | `rgba(0, 0, 0, 0.54)`
   * `--saved-request-detail-data-list-color` | Color of propery items | `rgba(0, 0, 0, 0.87)`,
   * `--saved-request-detail-action-button-color` | Color of action button | `--primary-color`
   * `--saved-request-detail-action-button-background-color` | Background color of action button | ``
   */
  class SavedRequestDetail extends
    Polymer.IronResizableBehavior(
    Object) {

    /**
     * Request object to render.
     */
    request: object|null|undefined;

    /**
     * Computed value, true if the request has been saved in the data store
     * whether it's saved or history.
     * Because request object can have `_id` generated before saving it to
     * the store this relays on checking both `_id` and `_rev`
     *    
     */
    readonly _isSaved: any;
    readonly projectModel: any;

    /**
     * Enables compatibility with Anypoint platform
     */
    compatibility: boolean|null|undefined;

    /**
     * Projects data associated with the request.
     */
    _projects: Array<object|null>|null;
    _titleTemplate(request: any): any;
    _addressTemplate(request: any): any;
    _descriptionTemplate(request: any): any;
    _timeTemplate(label: any, value: any, other: any): any;
    _driveTemplate(request: any): any;
    _projectsTemplate(): any;
    _deleteButtonTemplate(): any;
    _actionsTemplate(): any;
    _modelTemplate(): any;
    render(): any;
    connectedCallback(): void;
    disconnectedCallback(): void;
    firstUpdated(): void;

    /**
     * Sets project data when `request` object change.
     */
    _requestChanged(request: object|null): any;

    /**
     * Reads related project data.
     *
     * @param keys List of project IDs to read.
     */
    _readProjects(keys: Array<String|null>|null): Promise<any>|null;

    /**
     * Processes query response from the model.
     *
     * @param data The response
     * @param keys Requested keys
     * @returns Processed response or undefined.
     */
    _processProjectsResponse(data: Array<object|null>|null, keys: Array<String|null>|null): Array<object|null>|null|undefined;

    /**
     * Sends `navigate` event set to current read project.
     */
    _openProject(e: ClickEvent|null): void;

    /**
     * Sends non-bubbling `delete-request` event to the parent element to perform
     * delete action.
     */
    _deleteRequest(): void;

    /**
     * Sends non-bubbling `edit-request` event to the parent element to perform
     * edit action.
     */
    _editRequest(): void;
  }
}

declare global {

  interface HTMLElementTagNameMap {
    "saved-request-detail": UiElements.SavedRequestDetail;
  }
}


src/SavedRequestDetail.js
/**
@license
Copyright 2018 The Advanced REST client authors <arc@mulesoft.com>
Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html, css } from 'lit-element';
import { ArcResizableMixin } from '@advanced-rest-client/arc-resizable-mixin/arc-resizable-mixin.js';
import markdownStyles from '@advanced-rest-client/markdown-styles/markdown-styles.js';
import '@api-components/http-method-label/http-method-label.js';
import '@advanced-rest-client/date-time/date-time.js';
import { deleteIcon, edit } from '@advanced-rest-client/arc-icons/ArcIcons.js';
import '@anypoint-web-components/anypoint-button/anypoint-button.js';
import '@advanced-rest-client/arc-marked/arc-marked.js';
import '@anypoint-web-components/anypoint-chip/anypoint-chip.js';
import '@advanced-rest-client/arc-models/project-model.js';
/**
 * Details applet for saved request object.
 *
 * The applet doesn't support data edit. Element / app hosting this element
 * must handle events sent by this element and support edit action.
 *
 * ### Example
 *
 * ```html
 * <saved-request-detail request="{...}"></saved-request-detail>
 * ```
 *
 * ### Styling
 *
 * `<saved-request-detail>` provides the following custom properties and
 * mixins for styling:
 *
 * Custom property | Description | Default
 * ----------------|-------------|----------
 * `--saved-request-detail-color` | Color of the element | `inherit`
 * `--saved-request-detail-background-color` | Background color of the element | `inherit`
 * `--saved-request-detail-padding` | Padding of the element. | ``
 * `--saved-request-detail-description-max-width` | Max width of the description element | `700px`
 * `--saved-request-detail-description-color` | Color of the request description | `rgba(0, 0, 0, 0.64)`
 * `--saved-request-detail-action-icon-color` | Color of the icon in the action button | `rgba(0, 0, 0, 0.54)`
 * `--saved-request-detail-data-list-color` | Color of propery items | `rgba(0, 0, 0, 0.87)`,
 * `--saved-request-detail-action-button-color` | Color of action button | `--primary-color`
 * `--saved-request-detail-action-button-background-color` | Background color of action button | ``
 *
 * @customElement
 * @memberof UiElements
 * @demo demo/index.html
 * @appliesMixin Polymer.IronResizableBehavior
 */
export class SavedRequestDetail extends ArcResizableMixin(LitElement) {
  static get styles() {
    return [
      markdownStyles,
      css`
      :host {
        display: block;
        color: var(--saved-request-detail-color, inherit);
        background-color: var(--saved-request-detail-background-color, inherit);
        padding: var(--saved-request-detail-padding);
        box-sizing: border-box;
        font-size: var(--arc-font-body1-font-size);
        font-weight: var(--arc-font-body1-font-weight);
        line-height: var(--arc-font-body1-line-height);
      }

      h2 {
        font-size: var(--arc-font-headline-font-size);
        font-weight: var(--arc-font-headline-font-weight);
        letter-spacing: var(--arc-font-headline-letter-spacing);
        line-height: var(--arc-font-headline-line-height);
      }

      .address-detail {
        display: flex;
        flex-direction: row;
        align-items: center;
      }

      .url-label {
        font-size: 16px;
        white-space: var(--arc-font-nowrap-white-space);
        overflow: var(--arc-font-nowrap-overflow);
        text-overflow: var(--arc-font-nowrap-text-overflow);
      }

      http-method-label {
        font-size: 13px;
        margin-right: 12px;
        padding: 8px;
      }

      .description {
        max-width: var(--saved-request-detail-description-max-width, 700px);
        color: var(--saved-request-detail-description-color, rgba(0, 0, 0, 0.64));
      }

      .meta-row {
        display: flex;
        flex-direction: row;
        align-items: center;
        color: var(--saved-request-detail-data-list-color, rgba(0, 0, 0, 0.87));
        height: 40px;
      }

      .meta-row .label {
        width: 160px;
      }

      .meta-row .value {
        white-space: var(--arc-font-nowrap-white-space);
        overflow: var(--arc-font-nowrap-overflow);
        text-overflow: var(--arc-font-nowrap-text-overflow);
      }

      .actions {
        display: flex;
        flex-direction: row;
        justify-content: flex-end;
        margin-top: 20px;
      }

      .actions anypoint-button {
        padding-left: 12px;
        padding-right: 12px;
      }

      anypoint-button .icon {
        margin-right: 12px;
        fill: var(--saved-request-detail-action-icon-color, rgba(0, 0, 0, 0.54));
        display: inline-block;
        width: 24px;
        height: 24px;
      }

      arc-marked {
        margin-top: 20px;
        padding: 0;
      }

      .project-link {
        color: var(--arc-link-color, #00A1DF);
      }

      anypoint-chip {
        cursor: pointer;
      }

      anypoint-chip[disabled] {
        --anypoint-chip-border: 1px var(--error-color) solid;
      }
      `
    ];
  }

  _titleTemplate(request) {
    const title = request.name || 'Request details'
    return html`<h2 id="title">${title}</h2>`;
  }

  _addressTemplate(request) {
    if (!request.method) {
      return;
    }
    return html`<div class="address-detail">
      <http-method-label method="${request.method}"></http-method-label>
      <span class="url-label">${request.url || ''}</span>
    </div>`;
  }

  _descriptionTemplate(request) {
    if (!request.description) {
      return '';
    }
    return html`<arc-marked markdown="${request.description}" sanitize>
      <div class="markdown-html markdown-body description"></div>
    </arc-marked>`;
  }

  _timeTemplate(label, value, other) {
    if (!value || value === other) {
      return;
    }
    return html`<div class="meta-row">
      <div class="label">
        ${label}
      </div>
      <div class="value">
        <date-time
          .date="${value}"
          year="numeric"
          month="numeric"
          day="numeric"
          hour="numeric"
          minute="numeric"></date-time>
      </div>
    </div>`;
  }

  _driveTemplate(request) {
    if (!request.driveId) {
      return '';
    }
    return html`<div class="meta-row">
      <div class="label">
        Drive file
      </div>
      <div class="value">
        ${request.driveId}
      </div>
    </div>`;
  }

  _projectsTemplate() {
    const _projects = this._projects;
    if (!_projects || !_projects.length) {
      return;
    }
    const { compatibility } = this;
    return html`<div class="meta-row">
      <div class="label">
        Projects
      </div>
      <div class="value">
      ${_projects.map((item, index) => html`
        <anypoint-chip
          ?disabled="${item.missing}"
          title="Open project details"
          @click="${this._openProject}"
          data-index="${index}"
          data-action="open-project"
          ?compatibility="${compatibility}">${item.name}</anypoint-chip>`)}
      </div>
    </div>`;
  }

  _deleteButtonTemplate() {
    if (!this._isSaved) {
      return '';
    }
    return html`<anypoint-button
      @click="${this._deleteRequest}"
      data-action="delete-request"
      title="Removes request from the data store"
      aria-label="Activate to remove the request"
      ?compatibility="${this.compatibility}"
    >
      <span class="icon">${deleteIcon}</span>
      Delete
    </anypoint-button>`;
  }

  _actionsTemplate() {
    return html`
    ${this._deleteButtonTemplate()}
    <anypoint-button
      @click="${this._editRequest}"
      data-action="edit-request"
      title="Opens request editor"
      aria-label="Activate to edit request"
      ?compatibility="${this.compatibility}"
    >
      <span class="icon">${edit}</span>
      Edit
    </anypoint-button>
    `;
  }

  _modelTemplate() {
    return html`<project-model></project-model>`;
  }

  render() {
    const request = this.request || {};
    return html`
    ${this._modelTemplate()}
    ${this._titleTemplate(request)}
    ${this._addressTemplate(request)}
    ${this._descriptionTemplate(request)}
    ${this._timeTemplate('Created', request.created)}
    ${this._timeTemplate('Updated', request.updated, request.created)}
    ${this._driveTemplate(request)}
    ${this._projectsTemplate()}
    <div class="actions">
      ${this._actionsTemplate()}
    </div>`;
  }

  static get properties() {
    return {
      // Request object to render.
      request: { type: Object },
      /**
       * Enables compatibility with Anypoint platform
       */
      compatibility: { type: Boolean },
      /**
       * Projects data associated with the request.
       * @type {Array<Object>}
       */
      _projects: Array
    };
  }

  get request() {
    return this._request;
  }

  set request(value) {
    const old = this._request;
    if (old === value) {
      return;
    }
    this._request = value;
    this.requestUpdate('request', old);
    this._requestChanged(value);
  }
  /**
   * Computed value, true if the request has been saved in the data store
   * whether it's saved or history.
   * Because request object can have `_id` generated before saving it to
   * the store this relays on checking both `_id` and `_rev`
   */
  get _isSaved() {
    const { request } = this;
    if (!request) {
      return false;
    }
    return !!(request._id && request._rev);
  }

  get projectModel() {
    if (!this.__projectModel) {
      this.__projectModel = this.shadowRoot.querySelector('project-model');
    }
    return this.__projectModel;
  }

  connectedCallback() {
    if (super.connectedCallback) {
      super.connectedCallback();
    }
    this.setAttribute('role', 'dialog');
    this.setAttribute('aria-label', 'Request details dialog');
  }

  disconnectedCallback() {
    if (super.disconnectedCallback) {
      super.disconnectedCallback();
    }
    this.__projectModel = null;
  }

  firstUpdated() {
    if (this.request) {
      this._requestChanged(this.request);
    }
  }

  /**
   * Sets project data when `request` object change.
   *
   * @param {Object} request
   */
  async _requestChanged(request) {
    this._projects = undefined;
    if (!request) {
      return;
    }
    let keys;
    if (request.projects) {
      keys = request.projects;
    } else if (request.legacyProject) {
      keys = [request.legacyProject];
    }
    if (keys) {
      await this._readProjects(keys);
    }
  }
  /**
   * Reads related project data.
   * @param {Array<String>} keys List of project IDs to read.
   * @return {Promise}
   */
  async _readProjects(keys) {
    if (!keys || !keys.length) {
      this._projects = undefined;
      return;
    }
    const model = this.projectModel;
    if (!model) {
      // The element is initializing. Once upgraded it will re-query.
      return;
    }
    const data = await model.listProjects(keys);
    this._projects = this._processProjectsResponse(data, keys);
    setTimeout(() => this.notifyResize());
  }
  /**
   * Processes query response from the model.
   * @param {?Array<Object>} data The response
   * @param {Array<String>} keys Requested keys
   * @return {Array<Object>|undefined} Processed response or undefined.
   */
  _processProjectsResponse(data, keys) {
    if (!data || !data.length) {
      return;
    }
    return data.map((item, i) => {
      if (!item) {
        item = {
          missing: true,
          name: keys[i]
        };
      } else {
        item.missing = false;
      }
      return item;
    });
  }
  /**
   * Sends `navigate` event set to current read project.
   *
   * @param {ClickEvent} e
   */
  _openProject(e) {
    e.preventDefault();
    e.stopPropagation();
    const index = Number(e.currentTarget.dataset.index);
    const id = this._projects[index]._id;
    this.dispatchEvent(new CustomEvent('navigate', {
      bubbles: true,
      cancelable: true,
      composed: true,
      detail: {
        base: 'project',
        id
      }
    }));
  }
  /**
   * Sends non-bubbling `delete-request` event to the parent element to perform
   * delete action.
   */
  _deleteRequest() {
    this.dispatchEvent(new CustomEvent('delete-request', {
      detail: {
        item: this.request
      }
    }));
  }

  /**
   * Sends non-bubbling `edit-request` event to the parent element to perform
   * edit action.
   */
  _editRequest() {
    this.dispatchEvent(new CustomEvent('edit-request', {
      detail: {
        item: this.request
      }
    }));
  }
  /**
   * Fired when the user click on the "edit" action button.
   *
   * This event does not bubbles.
   *
   * @event edit-request
   * @param {Object} item The request object
   */
  /**
   * Fired when the user click on the "delete" action button.
   *
   * This event does not bubbles.
   *
   * @event delete-request
   * @param {Object} item The request object
   */
  /**
   * Fired when the user opens the project item.
   *
   * This event can be fired only if the request has a project.
   *
   * @event navigate
   * @param {String} base Always `project`
   * @param {String} id Project ID
   */
}
